File: 06. Backtracking/04. Permutations I/main.py
def backtrack(nums, used, current, result):
    if len(current) == len(nums):
        result.append(current[:])
        return
    
    for i in range(len(nums)):
        if used[i]:
            continue
        
        used[i] = True
        current.append(nums[i])
        backtrack(nums, used, current, result)
        current.pop()
        used[i] = False

def permute(nums):
    result = []
    backtrack(nums, [False] * len(nums), [], result)
    return result

def main():
    nums = [1, 2, 3]
    print("Permutations:", permute(nums))

# Run the main function
if __name__ == "__main__":
    main()

File: 06. Backtracking/04. Permutations I/main.js
function backtrack(nums, used, current, result) {
    if (current.length === nums.length) {
        result.push([...current]);
        return;
    }
    
    for (let i = 0; i < nums.length; i++) {
        if (used[i]) continue;
        
        used[i] = true;
        current.push(nums[i]);
        backtrack(nums, used, current, result);
        current.pop();
        used[i] = false;
    }
}

function permute(nums) {
    const result = [];
    backtrack(nums, Array(nums.length).fill(false), [], result);
    return result;
}

function main() {
    const nums = [1, 2, 3];
    console.log("Permutations:", permute(nums));
}

// Run the main function
main();
